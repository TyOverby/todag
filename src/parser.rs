use itertools::{put_back, PutBack};
use pulldown_cmark::{Event, Options, Parser, Tag};
use Event::*;

#[derive(Debug)]
pub struct Task<'a> {
    pub checked: bool,
    pub path: Vec<String>,
    pub name: Vec<Event<'a>>,
    pub description: Vec<Event<'a>>,
    pub nested: Vec<Task<'a>>,
}

#[derive(Debug)]
pub struct Header {
    depth: i32,
    text: String,
}

#[derive(Copy, Clone)]
enum TaskParseState {
    WaitingForTaskListMarker,
    ParsingTitle,
    ParsingDescription,
}

use TaskParseState::*;

fn parse_tasks<'a>(out: &mut Vec<Task<'a>>, paths: &[Header], parser: &mut PutBack<Parser<'a>>) {
    while let Some(event) = parser.next() {
        match event {
            Start(Tag::Item) => out.push(parse_task(paths, parser)),
            End(Tag::List(_)) => break,
            other => eprintln!("{:?} not implemented", other),
        }
    }
}

fn parse_task<'a>(paths: &[Header], parser: &mut PutBack<Parser<'a>>) -> Task<'a> {
    let mut checked = false;
    let mut name = Vec::new();
    let mut description = Vec::new();
    let path = paths.iter().map(|header| &header.text).cloned().collect();
    let mut state = WaitingForTaskListMarker;
    let mut nested = vec![];
    while let Some(event) = parser.next() {
        match (event, state) {
            (TaskListMarker(c), WaitingForTaskListMarker) => {
                checked = c;
                state = ParsingTitle;
            }
            (SoftBreak, ParsingTitle) => state = ParsingDescription,
            (event @ End(Tag::Item), ParsingTitle) => {
                parser.put_back(event);
                state = ParsingDescription;
            }
            (Start(Tag::List(_)), ParsingTitle) | 
            (Start(Tag::List(_)), ParsingDescription)=> {
                parse_tasks(&mut nested, paths, parser);
                state = ParsingDescription;
            }
            (event, ParsingTitle) => name.push(event),
            (event @ End(Tag::Item), ParsingDescription) => {
                parser.put_back(event);
                break;
            }
            (event, ParsingDescription) => description.push(event),
            _ => break,
        }
    }
    Task {
        path,
        name,
        checked,
        description,
        nested,
    }
}

fn parse_header(depth: i32, parser: &mut PutBack<Parser>) -> Header {
    let mut text = String::new();
    while let Some(event) = parser.next() {
        match event {
            Event::Text(t) => text.push_str(&t),
            Event::End(Tag::Header(_)) => break,
            _ => break,
        }
    }
    Header { depth, text }
}

pub fn extract(s: &str) -> (Vec<Task>, Vec<Header>) {
    let mut options = Options::empty();
    options.insert(Options::ENABLE_TASKLISTS);

    let mut parser = put_back(Parser::new_ext(s, options));
    let mut out = Vec::new();
    let mut path_stack: Vec<Header> = Vec::new();
    while let Some(event) = parser.next() {
        match event {
            Event::Start(Tag::Header(i)) => {
                let header = parse_header(i, &mut parser);
                while let Some(top_header) = path_stack.pop() {
                    if top_header.depth < header.depth {
                        path_stack.push(top_header);
                        break;
                    }
                }
                path_stack.push(header);
            }
            Event::Start(Tag::List(None)) => {
                parse_tasks(&mut out, &path_stack, &mut parser);
            }
            other => {
                eprintln!("{:?} not implemented", other);
            }
        }
    }

    (out, path_stack)
}
