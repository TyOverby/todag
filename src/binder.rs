use crate::parser::Task;
use pulldown_cmark::Event;
use std::collections::HashMap;

#[derive(Hash, Debug, Eq, PartialEq, Clone, Copy, PartialOrd, Ord)]
struct Id(usize);

struct IdGen {
    current: usize,
}

impl IdGen {
    fn new() -> IdGen {
        IdGen { current: 0 }
    }
    fn next(&mut self) -> Id {
        let n = self.current;
        self.current += 1;
        Id(n)
    }
}

#[derive(Debug)]
pub struct BoundTask<'a> {
    id: Id,
    name: Vec<Event<'a>>,
    description: Vec<Event<'a>>,
    depends_on: Vec<Id>,
}

impl<'a> BoundTask<'a> {
    fn from_task<'b>(id: Id, task: &Task<'b>) -> BoundTask<'b> {
        BoundTask {
            id,
            name: task.name.clone(),
            description: task.description.clone(),
            depends_on: vec![],
        }
    }
}

fn bind_chain<'a>(
    parent: &mut BoundTask<'a>,
    task: &Task<'a>,
    id_to_task: &mut HashMap<Id, BoundTask<'a>>,
    id_gen: &mut IdGen,
) {
    let id = id_gen.next();
    let mut bound_task = BoundTask::from_task(id, task);
    parent.depends_on.push(id);

    for child in &task.nested {
        bind_chain(&mut bound_task, child, id_to_task, id_gen);
    }

    id_to_task.insert(id, bound_task);
}

pub fn bind<'a>(tasks: &[Task<'a>]) -> Vec<BoundTask<'a>> {
    let mut id_gen = IdGen::new();
    let mut id_to_task = HashMap::new();
    for task in tasks {
        let id = id_gen.next();
        let mut bound_task = BoundTask::from_task(id, task);
        for child in &task.nested {
            bind_chain(&mut bound_task, child, &mut id_to_task, &mut id_gen);
        }
        id_to_task.insert(id, bound_task);
    }

   let mut all_tasks: Vec<_> = id_to_task.into_iter().map(|(_, a)|a).collect();
   all_tasks.sort_by_key(|task| task.id);
   all_tasks
}
