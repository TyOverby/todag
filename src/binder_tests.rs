use crate::binder::*;
use crate::parser::*;

fn run_bind<'a>(s: &'a str) -> Vec<BoundTask<'a>> {
    let (tasks, _headers) = extract(s);
    bind(&tasks[..])
}

#[test]
fn single_task() {
    let bound = run_bind(r#"
- [ ] this is the title 
  this is the description
"#);

    insta::assert_debug_snapshot!(bound);
}

#[test]
fn two_tasks_nested() {
    let bound = run_bind(r#"
- [ ] Title of A
  - [ ] Title of B
"#);

    insta::assert_debug_snapshot!(bound);
}

#[test]
fn two_tasks() {
    let bound = run_bind(r#"
- [ ] Title of A
- [ ] Title of B
"#);

    insta::assert_debug_snapshot!(bound);
}
