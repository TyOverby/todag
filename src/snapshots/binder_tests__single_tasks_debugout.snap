---
source: src/binder_tests.rs
expression: bound
---
[
    BoundTask {
        id: Id(
            0,
        ),
        name: [
            Text(
                Borrowed(
                    "this is the title ",
                ),
            ),
        ],
        description: [
            Text(
                Borrowed(
                    "this is the description",
                ),
            ),
        ],
        depends_on: [],
    },
]
